community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/WorkSync.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.transaction.command;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Turns multi-threaded unary work into single-threaded stack work.
+ * <p>
+ *     The technique used here is inspired in part both by the Flat Combining
+ *     concept from Hendler, Incze, Shavit & Tzafrir, and in part by the
+ *     wait-free linked queue design by Vyukov.
+ * </p>
+ * <p>
+ *     In a sense, this turns many small, presumably concurrent, pieces of work
+ *     into fewer, larger batches of work, that is then applied to the material
+ *     under synchronisation.
+ * </p>
+ * <p>
+ *     Obviously this only makes sense for that a) can be combined, and b)
+ *     where the performance improvements from batching effects is large enough
+ *     to overcome the overhead of collecting and batching up the work units.
+ * </p>
+ */
+public class WorkSync<Material, Work extends WorkSync.CombinableWork<Material,Work>>
```

Comment:
Excellent idea. I probably spiked the initial implementation and forgot to add tests.